skillmap/src/App.tsx

```diff
@@ -0,0 +1,294 @@
+/// <reference path="./lib/skillMap.d.ts" />
+
+import React from 'react';
+import { connect } from 'react-redux';
+
+import * as Promise from "bluebird";
+import store from "./store/store";
+
+import {
+    dispatchAddSkillMap,
+    dispatchClearSkillMaps,
+    dispatchSetPageTitle,
+    dispatchSetPageDescription,
+    dispatchSetPageInfoUrl,
+    dispatchSetUser,
+    dispatchSetPageSourceUrl
+} from './actions/dispatch';
+import { SkillMapState } from './store/reducer';
+import { HeaderBar } from './components/HeaderBar';
+import { Banner } from './components/Banner';
+import { AppModal } from './components/AppModal';
+import { SkillCarousel } from './components/SkillCarousel';
+
+import { parseSkillMap } from './lib/skillMapParser';
+import { parseHash, getMarkdownAsync, MarkdownSource, parseQuery, guidGen } from './lib/browserUtils';
+
+import './App.css';
+import { MakeCodeFrame } from './components/makecodeFrame';
+import { getUserStateAsync, saveUserStateAsync } from './lib/workspaceProvider';
+import { Unsubscribe } from 'redux';
+(window as any).Promise = Promise;
+
+interface AppProps {
+    skillMaps: { [key: string]: SkillMap };
+    activityOpen: boolean;
+    dispatchAddSkillMap: (map: SkillMap) => void;
+    dispatchClearSkillMaps: () => void;
+    dispatchSetPageTitle: (title: string) => void;
+    dispatchSetPageDescription: (description: string) => void;
+    dispatchSetPageInfoUrl: (infoUrl: string) => void;
+    dispatchSetUser: (user: UserState) => void;
+    dispatchSetPageSourceUrl: (url: string) => void;
+}
+
+interface AppState {
+    error?: string;
+}
+
+class AppImpl extends React.Component<AppProps, AppState> {
+    protected queryFlags: {[index: string]: string} = {};
+    protected unsubscribeChangeListener: Unsubscribe | undefined;
+    protected loadedUser: UserState | undefined;
+
+    constructor(props: any) {
+        super(props);
+        this.state = {};
+
+        window.addEventListener("hashchange", this.handleHashChange);
+    }
+
+    protected handleHashChange = (e: HashChangeEvent) => {
+        let hash = parseHash();
+        this.fetchAndParseSkillMaps(hash.cmd as MarkdownSource, hash.arg);
+
+        e.stopPropagation();
+        e.preventDefault();
+    }
+
+    protected async initLocalizationAsync() {
+        const theme = pxt.appTarget.appTheme;
+
+        const href = window.location.href;
+        let force = false;
+        let useLang: string | undefined = undefined;
+        if (/[&?]translate=1/.test(href) && !pxt.BrowserUtils.isIE()) {
```

Comment:
sidenote: I see this was just taken from another spot, but we're probably safe removing all the `pxt.BrowserUtils.isIE` references in the codebase, right? Seems fine to leave in here for now for consistency sake though~